File: app/utils/dawproject_python/project.py
from lxml import etree as ET
from classes.application import Application
from classes.transport import Transport
from classes.lane import Lane
from classes.arrangement import Arrangement
from classes.scene import Scene


class Project:
    CURRENT_VERSION = "1.0"

    def __init__(
        self,
        version=None,
        application=None,
        transport=None,
        structure=None,
        arrangement=None,
        scenes=None,
    ):
        self.version = version if version else self.CURRENT_VERSION
        self.application = application if application else Application()
        self.transport = transport
        self.structure = structure if structure else []
        self.arrangement = arrangement
        self.scenes = scenes if scenes else []

    def to_xml(self):
        root = ET.Element("Project", version=self.version)
        print(f"Creating XML for Project with version: {self.version}")

        # Correctly handling Application element with attributes
        app_elem = self.application.to_xml()
        root.append(app_elem)
        print("Added Application element.")

        if self.transport:
            transport_elem = ET.SubElement(root, "Transport")
            transport_elem.append(self.transport.to_xml())
            print("Added Transport element.")

        if self.structure:
            structure_elem = ET.SubElement(root, "Structure")
            for lane in self.structure:
                structure_elem.append(lane.to_xml())
            print("Added Structure element with lanes.")

        if self.arrangement:
            arrangement_elem = ET.SubElement(root, "Arrangement")
            arrangement_elem.append(self.arrangement.to_xml())
            print("Added Arrangement element.")

        if self.scenes:
            scenes_elem = ET.SubElement(root, "Scenes")
            for scene in self.scenes:
                scene_elem = ET.SubElement(scenes_elem, "Scene")
                scene_elem.append(scene.to_xml())
            print("Added Scenes element with scenes.")

        return root

    @classmethod
    def from_xml(cls, element):
        version = element.get("version", cls.CURRENT_VERSION)
        application = Application.from_xml(element.find("Application"))
        transport_elem = element.find("Transport")
        transport = (
            Transport.from_xml(transport_elem) if transport_elem is not None else None
        )

        structure_elem = element.find("Structure")
        structure = (
            [Lane.from_xml(lane) for lane in structure_elem]
            if structure_elem is not None
            else []
        )

        arrangement_elem = element.find("Arrangement")
        arrangement = (
            Arrangement.from_xml(arrangement_elem)
            if arrangement_elem is not None
            else None
        )

        scenes_elem = element.find("Scenes")
        scenes = (
            [Scene.from_xml(scene) for scene in scenes_elem]
            if scenes_elem is not None
            else []
        )

        return cls(version, application, transport, structure, arrangement, scenes)

File: app/utils/dawproject/project.ts
import { Application } from "./application";
import { Arrangement } from "./arrangement";
import { Lane } from "./lane";
import { LaneRegistry } from "./registry/laneRegistry";
import { Scene } from "./scene";
import { Transport } from "./transport";
import { IProject } from "./types";
import { Utility } from "./utility";
import { XmlObject } from "./XmlObject";

/** The main root element of the DAWPROJECT format. This is stored in the file project.xml file inside the container. */
export class Project extends XmlObject implements IProject {
  static CURRENT_VERSION = "1.0";

  /** Version of DAWPROJECT format this file was saved as. */
  version: string;
  /** Metadata (name/version) about the application that saved this file. */
  application: Application;
  /** Transport element containing playback parameters such as Tempo and Time-signature. */
  transport?: Transport;
  /** Track/Channel structure of this file. */
  structure: Lane[];
  /** The main Arrangement timeline of this file. */
  arrangement?: Arrangement;
  /** Clip Launcher scenes of this file. */
  scenes: Scene[];

  constructor(
    version?: string,
    application?: Application,
    transport?: Transport,
    structure?: Lane[],
    arrangement?: Arrangement,
    scenes?: Scene[]
  ) {
    super();
    this.version = version || Project.CURRENT_VERSION;
    this.application = application || new Application(); // Provide default values for required attributes
    this.transport = transport;
    this.structure = structure || [];
    this.arrangement = arrangement;
    this.scenes = scenes || [];
  }

  toXmlObject(): any {
    const obj: any = {
      Project: {},
    };

    // add required attribute
    Utility.addAttribute(obj.Project, "version", this, { required: true });

    if (this.application) {
      obj.Project.Application = this.application.toXmlObject().Application;
    }

    if (this.transport) {
      obj.Project.Transport = this.transport.toXmlObject().Transport;
    }

    // add children directly
    const groupedStructure = Utility.groupChildrenByTagName(this.structure);
    if (groupedStructure) {
      obj.Project.Structure = {
        ...groupedStructure,
      };
    }

    if (this.arrangement) {
      obj.Project.Arrangement = this.arrangement.toXmlObject().Arrangement;
    }

    if (this.scenes && this.scenes.length > 0) {
      obj.Project.Scenes = {
        Scene: this.scenes.map(scene => scene.toXmlObject().Scene),
      };
    }

    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    // validate and populate required attribute
    Utility.populateAttribute<string>(xmlObject, "version", this, {
      required: true,
    });

    if (xmlObject.Application) {
      this.application = new Application().fromXmlObject(xmlObject.Application);
    }

    if (xmlObject.Transport) {
      this.transport = new Transport().fromXmlObject(xmlObject.Transport);
    }

    const structure: Lane[] = [];
    if (xmlObject.Structure) {
      // Use the LaneRegistry to determine the correct subclass of Lane
      for (const tagName in xmlObject.Structure) {
        // Skip attributes (those starting with @_)
        if (tagName.startsWith("@_")) continue;

        const laneData = xmlObject.Structure[tagName];
        const laneArray = Array.isArray(laneData) ? laneData : [laneData];

        laneArray.forEach((laneObj: any) => {
          try {
            const laneInstance = LaneRegistry.createLaneFromXml(
              tagName,
              laneObj
            );
            if (laneInstance) {
              structure.push(laneInstance);
            } else {
              console.warn(
                `Skipping deserialization of unknown nested lane element in Structure: ${tagName}`
              );
            }
          } catch (e) {
            console.error(
              `Error deserializing nested lane element in Structure: ${tagName}`,
              e
            );
          }
        });
      }
    }
    this.structure = structure;

    if (xmlObject.Arrangement) {
      this.arrangement = new Arrangement().fromXmlObject(xmlObject.Arrangement);
    }

    const scenes: Scene[] = [];
    if (xmlObject.Scenes && xmlObject.Scenes.Scene) {
      const sceneArray = Array.isArray(xmlObject.Scenes.Scene)
        ? xmlObject.Scenes.Scene
        : [xmlObject.Scenes.Scene];

      sceneArray.forEach((sceneObj: any) => {
        scenes.push(new Scene().fromXmlObject(sceneObj));
      });
    }
    this.scenes = scenes;

    return this;
  }
}
